api/src/Validator/ContentNode/AssertSlotSupportedByParentValidator.php

```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace App\Validator\ContentNode;
+
+use App\Entity\ContentNode;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+use Symfony\Component\Validator\Exception\UnexpectedValueException;
+
+class AssertSlotSupportedByParentValidator extends ConstraintValidator {
+    public function __construct() {
+    }
+
+    public function validate($value, Constraint $constraint) {
+        if (!$constraint instanceof AssertSlotSupportedByParent) {
+            throw new UnexpectedTypeException($constraint, AssertSlotSupportedByParent::class);
+        }
+
+        $object = $this->context->getObject();
+        if (!$object instanceof ContentNode) {
+            throw new UnexpectedValueException($object, ContentNode::class);
+        }
+
+        $parent = $object->parent;
+        $supportedSlotNames = $parent?->getSupportedSlotNames() ?? [null];
```

Comment:
If the parent is null, then slot name must be null. The [null] captures this case.
getSupportedSlotNames() returns an [] which may be empty but never should be null or [null], ([null] would be possible by contract, but just give your slot a name)
If the array is empty, then the parent does not support any slot names, thus the relation of this contentNode to its parent is invalid. (And this error is also captured by the `[Assert\Type(type: SupportsContentNodeChildren::class)]`constraint).

But you have 2 error messages then.
Should i leave the validation of invalid parents to the Assert\Type constraint, or even remove it because this check also captures the other error without needing a type check?

Reply:
I see, thanks for the clarification. Maybe a comment on this line could help to distinguish the difference between the two cases `$supportedSlotNames == []` and `$parent == null` / `$supportedSlotNames == [null]`.

Two messages is better than none. But In my opinion, we could omit the type constraint if every time that the type constraint fails, another constraint also fails at the same time. I like the very clear validation message 'This parent does not support children, only content_nodes of type column_layout support children.' which we have on the type constraint so far, but we can also set a similarly descriptive message on the `AssertSlotSupportedByParent` constraint.

Now that I think about it, I would either remove the `SupportsContentNodeChildren` interface completely, or move the declaration of the `getSupportedSlotNames` method to that interface. What do you think about that?

Reply:
I have a response for that. I propose we discuss that at the next meeting.

Reply:
implemented with https://github.com/ecamp/ecamp3/pull/3126/commits/8c0ad3836952da37cc458a716caa155746ac9a65